File: 31-next-permutation/next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int b=0;
        int n= nums.size();
        for (int i=n-1 ; i>0 ; i--){
            if (nums[i-1]<nums[i]){
                b=i;
                break;
            }
        }

        if(b==0){
            reverse(nums.begin() , nums.end());
        }else{
            int key= b-1;
            int c=INT_MAX;
            int pos;
            for (int j= key+1 ; j<n; j++){
                if (nums[j]>nums[key] && nums[j]<c){
                    c= nums[j];
                    pos=j;
                }
            }
            swap(nums[key] , nums[pos]);
            sort(nums.begin()+key+1 , nums.end());
        }

        
    }
};

File: 31-next-permutation/next-permutation.py
def reverse_array(nums, start, end):
        while(start<end):
             nums[start], nums[end]= nums[end], nums[start]
             start+=1
             end-=1

        return nums

class Solution(object):
     



    def nextPermutation( self,nums):
            """
            :type nums: List[int]
            :rtype: None Do not return anything, modify nums in-place instead.
            """
            n= len (nums)
            pivot = -1
            

            for i in range (n-1 , 0, -1):
                if (nums[i-1]<nums[i]):
                    pivot= i-1
                    break

            print(pivot)

            if (pivot == -1):
                return reverse_array(nums, 0, n-1)

            for i in range (n-1 , -1, -1):
                if (nums[i]> nums[pivot]):
                    nums[i], nums[pivot]= nums[pivot], nums[i]
                    break

            nums= reverse_array(nums,pivot+1, n-1)
            return nums


        
        